### PR TITLE
Decompile code_800D2E30 and use struct in code_800A9F30

### PR DESCRIPTION
**Pull request by Synray**
Also matched a random function in z_camera while testing the permuter fix for pragma intrinsics.

### CODE REVIEW

File: src/code/z_camera.c [hunk 1]
```diff
@@ -22,9 +22,20 @@ typedef struct {
...
-// 47 lines
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_camera/func_800437F0.s")
-// had an attempt at this but didn't get very close
+f32 func_800437F0(f32 arg0, f32 arg1) {
+    f32 percent = .4f;
+    f32 ret = fabsf(arg1);
+
+    if (arg0 < ret) {
+        ret = 1;
+    } else if (ret < (float)arg0 * (1 - percent)) { //! float cast needed to match
```
Comment (krimtonz) [inline src/code/z_camera.c]: we use f32 over float
```suggestion
    } else if (ret < (f32)arg0 * (1 - percent)) { //! float cast needed to match
```

File: src/code/code_800A9F30.c [hunk 2]
```diff
@@ -22,19 +22,42 @@ void func_800A9F6C(f32 a, u8 b, u8 c, u8 d) {
...
+    if (1000000.0f < a) {
+        temp1 = 1000;
+    } else {
+        temp1 = sqrtf(a);
+    }
+
+    if (temp1 < 1000 && b != 0 && d != 0) {
+        temp2 = b - (temp1 * 255) / 1000;
+
+        for (i = 0; i != 0x40; i++) {
```
Comment (krimtonz) [inline src/code/code_800A9F30.c:44]: Does this match with i < 0x40? 
```suggestion
        for (i = 0; i < 0x40; i++) {
```

File: src/code/code_800D2E30.c [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        return;
+    }
+    D_8012DBB0 = arg0->unk_105;
+
+    if (arg0->unk_104 == 2) {
+        for (i = 0; i < 4; ++i) {
+            gPadMgr.pakType[i] = 0;
+        }
+
+        for (i = 0; i != 0x40; i++) {
```
Comment (krimtonz) [inline src/code/code_800D2E30.c:34]: Same as above
```suggestion
        for (i = 0; i < 0x40; i++) {
```

File: src/code/code_800D2E30.c [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            arg0->unk_C4[i] = 0;
+            arg0->unk_84[i] = 0;
+            arg0->unk_44[i] = 0;
+            arg0->unk_04[i] = 0;
+        }
+        arg0->unk_106 = arg0->unk_108 = arg0->unk_10A = arg0->unk_10B = arg0->unk_10C = arg0->unk_10D = 0;
+        arg0->unk_104 = 1;
+    }
+    if (arg0->unk_104 != 0) {
+        for (i = 0; i != 0x40; i++) {
```
Comment (krimtonz) [inline src/code/code_800D2E30.c:44]: Same as above
```suggestion
        for (i = 0; i < 0x40; i++) {
```

File: src/code/code_800D2E30.c [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                    if (arg0->unk_04[index] < arg0->unk_04[i]) {
+                        index = i;
+                        arg0->rumbleEnable[0] = (temp >= 0x100);
+                    }
+                }
+            }
+        }
+        if (arg0->unk_10A != 0) {
+            if (arg0->unk_10B > 0) {
+                arg0->unk_10B = arg0->unk_10B - 1;
```
Comment (krimtonz) [inline src/code/code_800D2E30.c:74]:
```suggestion
                arg0->unk_10B--;
```

File: src/code/code_800D2E30.c [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            } else {
+                temp = arg0->unk_10A - arg0->unk_10C;
+                if (temp > 0) {
+                    arg0->unk_10A = temp;
+                } else {
+                    arg0->unk_10A = 0;
+                }
+            }
+            temp = arg0->unk_10D + arg0->unk_10A;
```
Comment (krimtonz) [inline src/code/code_800D2E30.c:83]: Does this match? 
```suggestion
            arg0->unk_10D *= 2
```

File: src/code/code_800D2E30.c [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                arg0->unk_108 = 5;
+            }
+        } else {
+            arg0->unk_108 = 0;
+            if ((++arg0->unk_106) >= 0x1C21) {
+                arg0->unk_104 = 0;
+            }
+        }
+    } else {
+        for (i = 0; i != 0x40; i++) {
```
Comment (krimtonz) [inline src/code/code_800D2E30.c:109]: Same as above
```suggestion
        for (i = 0; i < 0x40; i++) {
```

File: src/code/z_camera.c [hunk 8]
```diff
@@ -22,9 +22,20 @@ typedef struct {
     Vec3s pos;
 } SplineData; // originally called SplinedatZ
 
-// 47 lines
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_camera/func_800437F0.s")
-// had an attempt at this but didn't get very close
+f32 func_800437F0(f32 arg0, f32 arg1) {
+    f32 percent = .4f;
```
Comment (krimtonz) [inline src/code/z_camera.c:26]:
```suggestion
    f32 percent = 0.4f;
```

File: src/code/z_camera.c [hunk 9]
```diff
@@ -22,9 +22,20 @@ typedef struct {
...
+    f32 ret = fabsf(arg1);
+
+    if (arg0 < ret) {
+        ret = 1;
+    } else if (ret < (float)arg0 * (1 - percent)) { //! float cast needed to match
+        ret = (SQ(arg1) * (1 - percent)) / SQ((1 - percent) * arg0);
+    } else {
+        ret = 1 - (.4f * SQ(arg0 - ret)) / SQ(.4f * arg0);
```
Comment (krimtonz) [inline src/code/z_camera.c:34]:
```suggestion
        ret = 1 - (0.4f * SQ(arg0 - ret)) / SQ(0.4f * arg0);
```

File: src/code/code_800D2E30.c [hunk 10]
```diff
@@ -0,0 +1,127 @@
+#include <ultra64.h>
+#include <global.h>
+
+extern u8 D_8012DBB0;
+extern PadMgr gPadMgr;
```
Comment (fig02) [inline src/code/code_800D2E30.c:5]: is this extern needed? we usually reference global variables directly (like gSaveContext) just fine. If it affects codegen then ignore me

Comment (Synray) [reply src/code/code_800D2E30.c:5]: Doesn't compile without it, I could add it to variables.h

Comment (fig02) [reply src/code/code_800D2E30.c:5]: yeah if its being used as a global variable then I say is appropriate. Didnt realize it wasnt there already

**Pull request merged**